### ISSUE TITLE
`Read::read_exact` with a capacity-initialized `Vec`

### ISSUE THREAD

**Issue by woodruffw** [trailofbits/dylint#1593]
The following code looks reasonable, and might be how someone would write a TLV parser:

```rust
let mut data = Vec::with_capacity(size);

some_io.read_exact(&mut data)?;
```

...but it actually reads `0` bytes and always returns `Ok(0)`, since `read_exact` checks `data.len()` instead of `data.capacity()`. As a result, `some_io` is not consumed (and is potentially left in a bad state, if the implementer doesn't support zero-sized reads?) and `data` is still empty.
